### ISSUE TITLE
Crash on Chrome at first intraction

### ISSUE THREAD

**Issue by carlopio** [TimDaub/wasm-synth#4]
Disabling the cache (or opening for the first time) + rapidly clicking on the keyboard will have everything crash with the following traces on the console: (see bottom)

Cool project/blog post though.

Carlo

main.js:320 Using default labels. Set labels explicitly via 'labels' in the options parameter
iS.parseArray_ @ main.js:320
iS.start_ @ main.js:320
iS.__init__ @ main.js:320
iS @ main.js:320
draw @ main.js:320
componentDidMount @ main.js:320
Yy @ main.js:7
t.unstable_runWithPriority @ main.js:7
tp @ main.js:7
jy @ main.js:7
ky @ main.js:7
xy @ main.js:7
cm @ main.js:7
(anonymous) @ main.js:7
Ty @ main.js:7
vm @ main.js:7
render @ main.js:7
(anonymous) @ main.js:320
main.js:7 TypeError: Cannot read property 'port' of null
    at pS.onNoteOff (main.js:320)
    at main.js:13
    at Array.forEach (<anonymous>)
    at t.handleNoteChanges (main.js:13)
    at t.value (main.js:13)
    at Yy (main.js:7)
    at t.unstable_runWithPriority (main.js:7)
    at tp (main.js:7)
    at jy (main.js:7)
    at ky (main.js:7)
xv @ main.js:7
Rv.n.callback @ main.js:7
Op @ main.js:7
Tp @ main.js:7
Yy @ main.js:7
t.unstable_runWithPriority @ main.js:7
tp @ main.js:7
jy @ main.js:7
ky @ main.js:7
(anonymous) @ main.js:7
t.unstable_runWithPriority @ main.js:7
tp @ main.js:7
op @ main.js:7
ip @ main.js:7
pu @ main.js:7
Rf @ main.js:7
main.js:320 Uncaught TypeError: Cannot read property 'port' of null
    at pS.onNoteOff (main.js:320)
    at main.js:13
    at Array.forEach (<anonymous>)
    at t.handleNoteChanges (main.js:13)
    at t.value (main.js:13)
    at Yy (main.js:7)
    at t.unstable_runWithPriority (main.js:7)
    at tp (main.js:7)
    at jy (main.js:7)
    at ky (main.js:7)
onNoteOff @ main.js:320
(anonymous) @ main.js:13
(anonymous) @ main.js:13
value @ main.js:13
Yy @ main.js:7
t.unstable_runWithPriority @ main.js:7
tp @ main.js:7
jy @ main.js:7
ky @ main.js:7
(anonymous) @ main.js:7
t.unstable_runWithPriority @ main.js:7
tp @ main.js:7
op @ main.js:7
ip @ main.js:7
pu @ main.js:7
Rf @ main.js:7
synth.js:17 Worklet launched successfully

**Comment by TimDaub** [comment]
You're correct. I've noticed this problem a couple of times already too. My suspicion is that it has to do with how the `AudioContext` and the WASM module are initialized, [this function](https://github.com/TimDaub/wasm-synth/blob/master/src/js/SynthAdapter.js#L42)

Note how the `AudioWorklet` node is being added. Note also that I've used [a trick](https://timdaub.github.io/2020/02/19/wasm-synth/#the-emscripten-build-pipeline) to automatically load WASM in the worklet.

On the first press of a button, the `init` function is called to initialize the `AudioContext`. 
I've looked into this quite a lot already and I haven't found a good approach, unfortunately.

I've looked into [Google's experiments](https://googlechromelabs.github.io/web-audio-samples/audio-worklet/design-pattern/wasm/) and they're using the SystemJS/module syntax to dynamically load the WASM into the worker. But notice how I argue against that in the blog post given that this module is not available on all major browsers.

IMO it's a problem of the browsers at this stage. I still could need help:

- Make it work for all major browsers (desktop and mobile)
- Use this amazing polyfill instead of the broken one from the Google ChromeLabs team: https://github.com/chrisguttandin/standardized-audio-context
- File issues on browser vendors repositories asking how they're planning to have worklets and js modules work together
- Find some short term hack that solves the problem so that I can demo WASM SYNTH

**Comment by carlopi** [comment]
Hi, double personality here (same as OP).
I then saw was already in your to-do list, then feel free to close it.

I haven't a clear picture, but I was thinking that one solutions to what seems to be a race on init could be:
-instead of running init not on the first button pressed, runs it when the page loads (via another __start__ of whatever)
-and then in the same function __start__ wait for init to execute (even a arbitrary time could work as a hack) and after that abilitate press of buttons.

I think using the better polyfill is probably the safest way, since it offers some more compatibility.
Other things that seems easy but might improve a lot:
-detect the browser isn't working, and explain to switch to Chrome
-add a link to the repo + blog post on the synth itself
-map the first half to the keyboard to (Q-O on the whites + 1-9 for the black keys) and the rghtmost half to (Z-M whites + A-K blacks), so it somehow works from the keyboard (sort of like an organ)

Might at some point find some free time and help out, it's a cool project, well done.

**Comment by TimDaub** [comment]
> -instead of running init not on the first button pressed, runs it when the page loads (via another **start** of whatever)

Yes I'd love to do that but in Chrome `AudioContext` can only be started when:

> The AudioContext was not allowed to start. It must be resumed (or created) after a user gesture on the page. <URL>

**Comment by carlopi** [comment]
Then maybe having a popup or some other kind of user interaction where you just do run init?
You could for example have a start screen/pop-up with a brief explanation of what's going on, and clicking there initialize the worklet.

**Comment by TimDaub** [comment]
OH I'M SURE MY USERS WILL LOVE THIS CARRRRRRLLL!

![image](https://user-images.githubusercontent.com/2758453/80864594-01ae6700-8c84-11ea-9022-d0b5a3845174.png)
